test/helpers/foundry/extract.ts

```diff
@@ -0,0 +1,211 @@
+ * @param binaries - The list of binaries to extract
+ * @param dir - The destination directory
+ * @param checksums - The checksums to verify the binaries against
+ * @returns The list of binaries extracted
+ */
+
+export async function extractFrom(
+  url: URL,
+  binaries: string[],
+  dir: string,
+  checksums: { algorithm: string; binaries: Record<Binary, string> } | null,
+) {
+  const extract = url.pathname.toLowerCase().endsWith(Extension.Tar)
+    ? extractFromTar
+    : extractFromZip;
+  // write all files to a temporary directory first, then rename to the final
+  // destination to avoid accidental partial extraction. We don't use
+  // `os.tmpdir` for this because `rename` will fail if the directories are on
+  // different file systems.
+  const tempDir = `${dir}.downloading`;
+  const rmOpts = { recursive: true, maxRetries: 3, force: true };
+  try {
+    // clean up any previous in-progress downloads
+    await rm(tempDir, rmOpts);
+    // make the temporary directory to extract the binaries to
+    await mkdir(tempDir, { recursive: true });
+    const downloads = await extract(
+      url,
+      binaries,
+      tempDir,
+      checksums?.algorithm,
+    );
+    ok(downloads.length === binaries.length, 'Failed to extract all binaries');
+
+    const paths: string[] = [];
+    for (const { path, binary, checksum } of downloads) {
+      if (checksums) {
+        say(`verifying checksum for ${binary}`);
+        const expected = checksums.binaries[binary as Binary];
+        if (checksum === expected) {
+          say(`checksum verified for ${binary}`);
+        } else {
+          throw new Error(
+            `checksum mismatch for ${binary}, expected ${expected}, got ${checksum}`,
+          );
+        }
+      }
+      // add the *final* path to the list of binaries
+      paths.push(join(dir, relative(tempDir, path)));
+    }
+
+    // everything has been extracted; move the files to their final destination
+    await rename(tempDir, dir);
+    // return the list of extracted binaries
+    return paths;
+  } catch (e) {
+    // if things fail for any reason try to clean up a bit. it is very important
+    // to not leave `dir` behind, as its existence is a signal that the binaries
+    // are installed.
+    await Promise.all([rm(tempDir, rmOpts), rm(dir, rmOpts)]).catch(noop);
+    throw e;
+  }
+}
+/**
+ * Extracts the binaries from a tar archive.
+ *
+ * @param url - The URL of the archive to extract the binaries from
+ * @param binaries - The list of binaries to extract
+ * @param dir - The destination directory
+ * @param checksumAlgorithm - The checksum algorithm to use
+ * @returns The list of binaries extracted
+ */
+
+async function extractFromTar(
+  url: URL,
+  binaries: string[],
+  dir: string,
+  checksumAlgorithm?: string,
+) {
+  const downloads: {
+    path: string;
+    binary: string;
+    checksum?: string;
+  }[] = [];
+  await pipeline(
+    startDownload(url),
+    extractTar(
```

Comment:
We can use `extractTar` to only extract the binaries we asked for. So we do. Zoom zoom.